File: array_n_hashing/rust/two_sum.rs
use std::collections::HashMap;

fn main() {

    let nums: Vec<i32> = vec![3,4,5,6]; // target 7


}

fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut map: HashMap<i32, usize> = HashMap::new();

    for (i, &num) in nums.iter().enumerate() {
        let diff = target - num;

        match map.get(&diff) {
            Some(&j) => return vec![j as i32, i as i32],
            None => {}
        }

        map.insert(num, i);
    }

    vec![]
}

File: array_n_hashing/cpp/two_sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

std::vector<int> twoSum(std::vector<int> &nums, int target) {
    std::unordered_map<int, int> map;

    int diff;

    for (int i = 0; i < nums.size(); i++) {
        map[nums[i]] = i;
    }

    for (int i = 0; i < nums.size(); i++) {
        diff = target - nums[i];
        if (map.find(diff) != map.end() && map[diff] != i) {
            return {i, map[diff]};
        }
    }

    return {};
}